src/vm/jithost.cpp

```diff
@@ -0,0 +1,204 @@
+// to be covered by pooling done by the default malloc.
+//
+// - Keep up to some limit worth of memory, with loose affinization of memory blocks to threads.
+// - On finalizer thread, release the extra memory that was not used recently.
+//
+
+struct Slab
+{
+    Slab * pNext;
+    size_t size;
+    Thread* affinity;
+};
+
+static CrstStatic s_JitSlabAllocatorCrst;
+static Slab* s_pCurrentCachedList;
+static Slab* s_pPreviousCachedList;
+static size_t s_TotalCached;
+static DWORD s_lastFlush = 0;
+
+void* JitHost::allocateSlab(size_t size, size_t* pActualSize)
+{
+    size = max(size, sizeof(Slab));
+
+    Thread* pCurrentThread = GetThread();
+    if (s_pCurrentCachedList != NULL || s_pPreviousCachedList != NULL)
+    {
+        CrstHolder lock(&s_JitSlabAllocatorCrst);
+        Slab** ppCandidate = NULL;
+
+        for (Slab ** ppList = &s_pCurrentCachedList; *ppList != NULL; ppList = &(*ppList)->pNext)
+        {
+            Slab* p = *ppList;
+            if (p->size >= size && p->size <= 4 * size) // Avoid wasting more than 4x memory
+            {
+                ppCandidate = ppList;
+                if (p->affinity == pCurrentThread)
+                    break;
+            }
+        }
+
+        if (ppCandidate == NULL)
+        {
+            for (Slab ** ppList = &s_pPreviousCachedList; *ppList != NULL; ppList = &(*ppList)->pNext)
+            {
+                Slab* p = *ppList;
+                if (p->size == size) // Allocation from previous list requires exact match
+                {
+                    ppCandidate = ppList;
+                    if (p->affinity == pCurrentThread)
+                        break;
+                }
+            }
+        }
+
+        if (ppCandidate != NULL)
+        {
+            Slab* p = *ppCandidate;
+            *ppCandidate = p->pNext;
+
+            s_TotalCached -= p->size;
+            *pActualSize = p->size;
+
+            return p;
+        }
+    }
+
+    *pActualSize = size;
+    return new (nothrow) BYTE[size];
+}
+
+void JitHost::freeSlab(void* slab, size_t actualSize)
+{
+    _ASSERTE(actualSize >= sizeof(Slab));
+
+    if (actualSize < 0x100000) // Do not cache blocks that are more than 1MB
+    {
+        CrstHolder lock(&s_JitSlabAllocatorCrst);
+
+        if (s_TotalCached < 0x1000000) // Do not cache more than 16MB
+        {
+            s_TotalCached += actualSize;
+
+            Slab* pSlab = (Slab*)slab;
+            pSlab->size = actualSize;
+            pSlab->affinity = GetThread();
+            pSlab->pNext = s_pCurrentCachedList;
+            s_pCurrentCachedList = pSlab;
+            return;
+        }
+    }
+
+    delete [] (BYTE*)slab;
+}
+
+void JitHost::Init()
+{
+    s_JitSlabAllocatorCrst.Init(CrstLeafLock);
+}
+
+void JitHost::Reclaim()
+{
+    if (s_pCurrentCachedList != NULL || s_pPreviousCachedList != NULL)
+    {
+        DWORD ticks = ::GetTickCount();
```

Comment:
`GetTickCount64`?

Reply:
GetTickCount64 is slower on 32-bit. I think GetTickCount works just fine here.